### PR TITLE
Add chart resource delete methods

### PR DESCRIPTION
**Pull request by indyfree**
Towards https://github.com/giantswarm/giantswarm/issues/1902.

Implements the chart resource delete methods corresponding to #37.

### CODE REVIEW

File: service/chartconfig/v1/resource/chart/delete.go [hunk 1]
```diff
@@ -2,14 +2,70 @@ package chart
...
+		return nil, microerror.Mask(err)
+	}
+	desiredChartState, err := toChartState(desiredState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("finding out if the %s chart has to be deleted", desiredChartState.ChartName))
+
+	if currentChartState.ChartName != "" && currentChartState.ChartName == desiredChartState.ChartName {
```
Comment (indyfree) [inline service/chartconfig/v1/resource/chart/delete.go]: No sure, whether we delete by `ChartName` or should use `ReleaseName` here as the helm client does.

Comment (xh3b4sd) [reply service/chartconfig/v1/resource/chart/delete.go]: IMO this kind of condition is not very cool for our resource implementations. I saw this mostly in the aws-operator and in some places, like the cloudformation resource it was necessary to do it this way, but in most cases we only have to check if the current state is not nil or something. It does not really add up for me to check some arbitrary property of the state object. Like, why should a name be there but not some other property? Why don't we check the other properties? And so on.

Comment (indyfree) [reply service/chartconfig/v1/resource/chart/delete.go]: Alright will change that. :+1: 
Just out of interest: How do we now we are deleting the right chart? Thought we need `currentChartState.ChartName == desiredChartState.ChartName` to check that.

File: service/chartconfig/v1/resource/chart/delete.go [hunk 2]
```diff
@@ -2,14 +2,71 @@ package chart
...
+		return nil, microerror.Mask(err)
+	}
+	desiredChartState, err := toChartState(desiredState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("finding out if the %s chart has to be deleted", desiredChartState.ChartName))
+
+	if !reflect.DeepEqual(currentChartState, ChartState{}) && currentChartState.ChartName == desiredChartState.ChartName {
```
Comment (indyfree) [inline service/chartconfig/v1/resource/chart/delete.go]: DeepEqual cause `ChartState` is a struct with non-comparable fields

Comment (xh3b4sd) [reply service/chartconfig/v1/resource/chart/delete.go]: That is still weird. We return `nil` below and a pointer of the desired state. The chart state though seems to be a value. We should deal with just the same data structure and representation across a resource implementation. So the chart state here should be a pointer as well and then we can simply check if it is nil like I suggested before.

Comment (xh3b4sd) [reply service/chartconfig/v1/resource/chart/delete.go]: No blocker. Go ahead. Lets iterate here.

**Comment by indyfree** [comment]
Changed checking if all fields are equal, instead of only `chartName`, before deletion.
This allows for installing multiple instances of the same chart, as discussed here:

https://gigantic.slack.com/archives/C76JX6YLQ/p1522790741000480.

If there are no objections I am gonna merge.

**Comment by rossf7** [comment]
@indyfree No blocker this is cool. Let's go with this.

In general I agree with Tim about returning pointers because it simplifies the error handling. But this can be done separately.

**Pull request merged**